src/values/AbstractValue.js

```diff
@@ -245,31 +245,31 @@ export default class AbstractValue extends Value {
   mightBeNull(): boolean {
     let valueType = this.getType();
     if (valueType === NullValue) return true;
-    if (valueType !== Value) return false;
+    if (valueType !== PrimitiveValue && valueType !== Value) return false;
     if (this.values.isTop()) return true;
     return this.values.includesValueOfType(NullValue);
   }
 
   mightNotBeNull(): boolean {
     let valueType = this.getType();
     if (valueType === NullValue) return false;
-    if (valueType !== Value) return true;
+    if (valueType !== PrimitiveValue && valueType !== Value) return true;
     if (this.values.isTop()) return true;
     return this.values.includesValueNotOfType(NullValue);
   }
 
   mightBeNumber(): boolean {
     let valueType = this.getType();
-    if (valueType === NumberValue) return true;
-    if (valueType !== Value) return false;
+    if (Value.isTypeCompatibleWith(valueType, NumberValue)) return true;
+    if (valueType !== PrimitiveValue && valueType !== Value) return false;
     if (this.values.isTop()) return true;
     return this.values.includesValueOfType(NumberValue);
   }
 
   mightNotBeNumber(): boolean {
     let valueType = this.getType();
-    if (valueType === NumberValue) return false;
-    if (valueType !== Value) return true;
+    if (Value.isTypeCompatibleWith(valueType, NumberValue)) return false;
```

Comment:
If we ever introduce an IntegralValue as a subtype of NumberValue, this kind of thing will be necessary, so may as well do it now while touching this code.